src/python/pants/engine/legacy/graph.py

```diff
@@ -360,24 +361,45 @@ def transitive_dependents_of_addresses(self, addresses):
         return result
 
 
-@dataclass(frozen=True)
+@frozen_after_init
+@dataclass
 class HydratedTarget:
     """A wrapper for a fully hydrated TargetAdaptor object.
 
+    Why have this type if all information is stored in the underlying TargetAdaptor? We need it for
+    the type-driven graph to work properly. A HydratedTarget is a generic wrapper around different
+    target classes, whereas a TargetAdaptor has different types like PythonTestsAdaptor and
+    PythonLibraryAdaptor. We need to be able to work with targets both generically and depending
+    on their target type (i.e. unions).
+
     Transitive graph walks collect ordered sets of TransitiveHydratedTargets which involve a huge
-    amount of hashing: we implement eq/hash via direct usage of an Address field to speed that up.
+    amount of hashing: we implement eq/hash via a private Address field to speed that up.
     """
 
-    address: Address
     adaptor: TargetAdaptor
-    dependencies: Tuple[Address, ...]
 
-    @property
-    def addresses(self) -> Tuple:
-        return self.dependencies
+    def __init__(self, adaptor: TargetAdaptor) -> None:
+        self.adaptor = adaptor
+        # This field is set for efficient lookup of the address in `__hash__` and `__eq__` during
+        # graph walks. Directly accessing this field, rather than using `adaptor.address` is about
+        # 10x faster. However, rules should still use `adaptor.address` for clarity and because
+        # they do not need to access the address as frequently, so it would be an over-optimization
+        # (we're talking about 0.000_000_1 vs. 0.000_001 seconds here).
```

Comment:
Let's use scientific notation here, it's easier to reason about orders of magnitude. I.e., 10^-7 vs 10^-6.